bocadillo/auth.py

```diff
@@ -0,0 +1,98 @@
+import inspect
+import typing
+
+from starlette.authentication import (
+    AuthCredentials,
+    AuthenticationBackend,
+    AuthenticationError,
+    BaseUser,
+)
+from starlette.middleware.authentication import AuthenticationMiddleware
+from starlette.requests import HTTPConnection
+from starlette.responses import JSONResponse
+
+from .config import settings
+from .errors import HTTPError
+from .plugins import plugin
+
+AuthResult = typing.Optional[typing.Tuple[AuthCredentials, BaseUser]]
+
+
+@plugin
+def use_authentication(app: "App"):
+    backend = MultiAuth()
+    if not backend.backends:
+        return
+    app.add_asgi_middleware(AuthMiddleware, backend=backend)
+
+
+async def authenticate(conn: HTTPConnection, **kwargs) -> AuthResult:
+    backend = MultiAuth()
+    return await backend.authenticate(conn, **kwargs)
+
+
+class AuthMiddleware(AuthenticationMiddleware):
+    @staticmethod
+    def default_on_error(conn: HTTPConnection, exc: AuthenticationError):
+        return JSONResponse(
+            HTTPError(401, detail=str(exc)).as_json(), status_code=401
+        )
+
+
+def _ensure_instance(obj: typing.Any, *args, **kwargs) -> typing.Any:
+    return obj(*args, **kwargs) if inspect.isclass(obj) else obj
+
+
+class MultiAuth(AuthenticationBackend):
```

Comment:
@tomchristie I noticed I can easily refactor `MultiAuth` and the general-purpose `authenticate()` to be framework-agnostic, so I'm wondering whether this and `ModelAuth` could be eligible to land in Starlette?

I think there's a way to build generic-but-extensible auth for the async ecosystem. Feel free to go through https://github.com/bocadilloproject/bocadillo/issues/100#issuecomment-487315991 and this PR's description. :-)